### PR TITLE
[Reland] [13/N] Enable clang-tidy on headers of torch/csrc

### PR DESCRIPTION
**Pull request by cyyever**
Reland of #116560 and fixes the issued reported by #116695

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @izaitsevfb

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Thanks!

File: torch/csrc/Exceptions.h [hunk 1]
```diff
@@ -373,15 +371,14 @@ using Arg = typename invoke_traits<Func>::template arg<i>::type;
...
+    // NOLINTNEXTLINE(cppcoreguidelines-missing-std-forward)
     Func&& f,
     std::index_sequence<Is...>,
     bool release_gil) {
-  using result_type = typename invoke_traits<Func>::result_type;
   namespace py = pybind11;
 
   // f=f is needed to handle function references on older compilers
-  return [f = std::forward<Func>(f),
-          release_gil](Arg<Func, Is>... args) -> result_type {
```
Comment (Skylion007) [inline torch/csrc/Exceptions.h:384]: Why was result_type removed from this lambda?

Comment (cyyever) [reply torch/csrc/Exceptions.h:384]: I tried to reduce the dependency to C++17.h by using auto type deduction.

Comment (Skylion007) [reply torch/csrc/Exceptions.h:384]: Doesn't C++17  have any native APIs to do this? Presumably we can remove that dependency if we use the stdlib instead? Or does that cause issues on certain compilers?

**Pull request closed without merging**